### PR TITLE
[louist3ng] IP

### PR DESCRIPTION
**Pull request by louist3ng**
# 🤖 Jason Chatbot

Jason is a **Java-based CLI chatbot** (inspired by Duke) that helps users manage their tasks efficiently.  

> “Your personal task manager, right in the terminal.”

---

## ✨ Features
- Add tasks: todo, deadlines, events  
- Mark/unmark tasks as done  
- Delete tasks from the list  
- Save/load tasks from disk  
- Robust error handling with custom exceptions  

---

## 🚀 How It Works (Step by Step)
1. User enters a command through the CLI  
2. `CommandParser` interprets the input  
3. A corresponding `Command` object is executed  
4. `TaskList` and `Storage` are updated accordingly  

---

## 🖥️ Example Usage

```java
// Add a todo task
todo read CS2103T lecture notes

// Mark task 1 as done
mark 1

```

## ✅ To-Do List
- [x] Implement core commands (`todo`, `deadline`, `event`)  
- [ ] Support marking/unmarking tasks  
- [x] Persistent storage with `Storage` class  
- [ ] Implement `find` command  
- [ ] Enhance exception handling and messages  
- [ ] Add JavaFX GUI for better user experience  
- [ ] Write additional JUnit tests for edge cases  

---

## 📎 Useful Links
- [📘 Project README](https://github.com/louist3ng/ip/blob/master/README.md)  
- [📂 GitHub Repo (louist3ng/ip)](https://github.com/louist3ng/ip)

### CODE REVIEW

File: src/main/java/jason/app/Jason.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
+package jason.app;
```
Comment (Nelsonweena) [inline src/main/java/jason/app/Jason.java:1]: Good! noticed all files are in a package that is in lower case

File: src/main/java/jason/model/Task.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+     * Constructs a Task with the given description.
+     *
+     * @param description The description of the task.
+     */
+    public Task(String description) {
+        this.description = description;
+        isDone = false;
+    }
+
+    public String getDescription() {
```
Comment (Nelsonweena) [inline src/main/java/jason/model/Task.java:21]: Good! methods are in camelCase

File: src/main/java/jason/app/Jason.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+import jason.model.TaskList;
+import jason.model.Todo;
+import jason.parser.DateTimeUtil;
+import jason.parser.Parser;
+import jason.storage.Storage;
+import jason.ui.Ui;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.util.List;
+
```
Comment (Nelsonweena) [inline src/main/java/jason/app/Jason.java:14]: Might need some java documentation for this file 🤔

File: src/main/java/jason/app/Jason.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    public void run() {
+        loadNow();
+        ui.intro();
+
+        boolean exit = false;
+        while (!exit) {
+            String input = ui.readInput();
+            String cmd = Parser.taskCommand(input);
+            String contents = Parser.taskContent(input);
+
```
Comment (Nelsonweena) [inline src/main/java/jason/app/Jason.java]: might want to put these in the parser class instead.

File: src/main/java/jason/model/Todo.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package jason.model;
+
+/**
+ * Represents a todo task.
+ */
+public class Todo extends Task {
+
+    /**
```
Comment (Nelsonweena) [inline src/main/java/jason/model/Todo.java:8]: Good! apart from the Jason file, there is consistent use of java docs 👍

File: src/main/java/jason/ui/Ui.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    public void bye() {
+        println(LINE);
+        println("Bye. Hope to see you again soon!");
+        println(LINE);
+    }
+
+    /**
+     * Reads a line of input from the user.
+     * @return the input string
+     */
```
Comment (Nelsonweena) [inline src/main/java/jason/ui/Ui.java]: think this method should be under jason instead. 🤔

File: src/main/java/jason/ui/Ui.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        return in.nextLine();
+    }
+
+    /**
+<<<<<<< HEAD
+     * Call at the very end of main loop
+=======
+     * Call at the very end of main loop.
+>>>>>>> A-CodingStandard
+     */
```
Comment (Nelsonweena) [inline src/main/java/jason/ui/Ui.java]: think this method should be under jason instead. 🤔

File: src/main/java/jason/exception/ParentException.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package jason.exception;
+
+/**
+ * Base class for all custom exceptions in the application.
+ */
+public class ParentException extends RuntimeException {
```
Comment (denz-denz) [inline src/main/java/jason/exception/ParentException.java:6]: Maybe a more suitable name might be JasonException but not a coding violation

File: src/main/java/jason/model/Deadline.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
+package jason.model;
+
+import jason.parser.DateTimeUtil;
+import java.time.LocalDateTime;
+
+public class Deadline extends Task {
```
Comment (denz-denz) [inline src/main/java/jason/model/Deadline.java:9]: Could maybe add a java doc header above the class

File: src/main/java/jason/app/Jason.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+import jason.model.Task;
+import jason.model.TaskList;
+import jason.model.Todo;
+import jason.parser.DateTimeUtil;
+import jason.parser.Parser;
+import jason.storage.Storage;
+import jason.ui.Ui;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.util.List;
```
Comment (denz-denz) [inline src/main/java/jason/app/Jason.java:13]: importing java statements should be above import Jason statements

File: src/main/java/jason/model/Event.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    @Override
+    public String getDescription() {
+        return (isDone ? "[E][X] " : "[E][ ] ")
+            + description + " (from: " + from + " to: " + to + ")";
+    }
+
+    @Override
+    public String toStorageString() {
+        // E | 0 | description | 2019-12-02T18:00 | 2019-12-02T20:00
+        return "E | " + (isDone ? "1" : "0") + " | " + description
```
Comment (denz-denz) [inline src/main/java/jason/model/Event.java:44]: Nice formatting of + in new line

File: src/main/java/jason/model/Task.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        isDone = false;
+        return this;
+    }
+
+    /**
+     * Returns whether the task is done.
+     *
+     * @return True if the task is done, false otherwise.
+     */
+    public boolean isDone() {
```
Comment (denz-denz) [inline src/main/java/jason/model/Task.java:50]: good choice of method name which makes it very intuitive

File: src/main/java/jason/ui/Ui.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+    public void close() {
+        in.close();
+    }
+
+    /**
+     * Prints a message to the standard output.
+     * @param msg the message to print
+     */
+    public void println(String msg) {
```
Comment (denz-denz) [inline src/main/java/jason/ui/Ui.java]: maybe use a different name, else might be confused with System.out.println. Perhaps like showMessage()

File: src/main/java/jason/ui/Ui.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+        System.out.println(msg);
+    }
+
+    /**
+     * Prints an error message to the standard error output.
+     * @param msg the error message to print
+     */
+    public void error(String msg) {
```
Comment (denz-denz) [inline src/main/java/jason/ui/Ui.java]: Might be better to be named as a verb like showError()

File: src/main/java/jason/storage/Storage.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    public Storage(String relativePath) {
+        this.file = Paths.get(relativePath);
+    }
+
+    /**
+     * Ensures that the parent directory for the storage file exists.
+     * @throws IOException if an I/O error occurs
+     */
+    private void ensureDir() throws IOException {
```
Comment (denz-denz) [inline src/main/java/jason/storage/Storage.java:32]: I like the choice of method name here ✅

**Review by denz-denz — COMMENTED** [review]
Overall, other than some coding violations, very clean use of OOP. Might want to put some heavy lifting in main into the parser class
